File: src/main/python/movie-stats.py
#!/usr/bin/python
#  -*- coding:utf-8 -*-

from pyspark import SparkContext
from pyspark.mllib.feature import Normalizer
import matplotlib as mpl
mpl.use('Agg')
import matplotlib.pyplot as plt
import numpy as np
import datetime

sc = SparkContext()
# 读取数据集
movie_data = sc.textFile("hdfs://PATH/ml-100k/u.item")
# 输出第1行数据
print(movie_data.first())
# 输出前k行数据
print(movie_data.take(10))

# 统计电影数量
num_movies = movie_data.count()
# Movies: 1682
print("Movies: %d" % num_movies)

# 转换年份函数
def convert_year(x):
    try:
        return int(x[-4:])
    except:
        return 1900 # 若数据缺失年份则将其年份设为1900

movie_fields = movie_data.map(lambda line: line.split("|"))
# 提取出电影的年份信息
years = movie_fields.map(lambda fields: fields[2]).map(lambda x: convert_year(x))
# 过滤掉1900年(即未记录年份)的电影数据
years_filtered = years.filter(lambda x: x != 1900)
# 计算得到电影的年龄
today = datetime.date.today()
this_year = today.year
movie_ages = years_filtered.map(lambda year: this_year - year).countByValue()
values = movie_ages.values()
bins = movie_ages.keys()
# bins: 区间数
# normed=True: 正则化直方图
plt.hist(values, bins=bins, color='lightblue', normed=True)
#fig = plt.gcf()
#fig.set_size_inches(16,10)

# 年份集合
years_pre_processed = movie_fields.map(lambda fields: fields[2]).map(lambda x: convert_year(x)).collect()
years_pre_processed_array = np.array(years_pre_processed)
# 非1900年的全部年份的平均值
mean_year = np.mean(years_pre_processed_array[years_pre_processed_array!=1900])
# 非1900年的全部年份的中位数值
median_year = np.median(years_pre_processed_array[years_pre_processed_array!=1900])
# np.where(condition): 输出满足条件的元素的坐标
index_bad_data = np.where(years_pre_processed_array==1900)[0][0]
print("index_bad_data:")
print(index_bad_data)
years_pre_processed_array[index_bad_data] = median_year
# Mean year of release: 1989
print("Mean year of release: %d" % mean_year)
# Median year of release: 1995
print("Median year of release: %d" % median_year)
# Index of '1900' after assigning median: []
print("Index of '1900' after assigning median: %s" % np.where(years_pre_processed_array == 1900)[0])

def extract_title(raw):
    import re
    # 该表达式找寻括号之间的非单词（数字）
    grps = re.search("\((\w+)\)", raw)
    if grps:
        # 只选取标题部分，并删除末尾的空白字符
        return raw[:grps.start()].strip()
    else:
        return raw

raw_titles = movie_fields.map(lambda fields: fields[1])
for raw_title in raw_titles.take(5):
    print extract_title(raw_title)

movie_titles = raw_titles.map(lambda m: extract_title(m))
# 下面用简单空白分词法将标题分词为词
title_terms = movie_titles.map(lambda t: t.split(" "))
print title_terms.take(5)

# 下面取回所有可能的词，以便构建一个词到序号的映射字典
all_terms = title_terms.flatMap(lambda x: x).distinct().collect()
# 创建一个新的字典来保存词，并分配k之1序号
idx = 0
all_terms_dict = {}
for term in all_terms:
    all_terms_dict[term] = idx
    idx +=1

# Total number of terms: 2645
print "Total number of terms: %d" % len(all_terms_dict)
# Index of term 'Dead': 147
print "Index of term 'Dead': %d" % all_terms_dict['Dead']
# Index of term 'Rooms': 1963
print "Index of term 'Rooms': %d" % all_terms_dict['Rooms']

all_terms_dict2 = title_terms.flatMap(lambda x: x).distinct().zipWithIndex().collectAsMap()
# Index of term 'Dead': 147
print "Index of term 'Dead': %d" % all_terms_dict2['Dead']
# Index of term 'Rooms': 1963
print "Index of term 'Rooms': %d" % all_terms_dict2['Rooms']

# 该函数输入一个词列表，并用k之1编码类似的方式将其编码为一个scipy稀疏向量
def create_vector(terms, term_dict):
    from scipy import sparse as sp
    num_terms = len(term_dict)
    x = sp.csc_matrix((1, num_terms))
    for t in terms:
        if t in term_dict:
            idx = term_dict[t]
            x[0, idx] = 1
    return x

all_terms_bcast = sc.broadcast(all_terms_dict)
term_vectors = title_terms.map(lambda terms: create_vector(terms, all_terms_bcast.value))
term_vectors.take(5)

# 将随机种子的值设为42，以保证每次运行的结果相同
np.random.seed(42)
x = np.random.randn(10)
norm_x_2 = np.linalg.norm(x)
normalized_x = x / norm_x_2
print "x:\n%s" % x
print "2-Norm of x: %2.4f" % norm_x_2
print "Normalized x:\n%s" % normalized_x
print "2-Norm of normalized_x: %2.4f" % np.linalg.norm(normalized_x)

normalizer = Normalizer()
vector = sc.parallelize([x])
normalized_x_mllib = normalizer.transform(vector).first().toArray()
print "x:\n%s" % x
print "2-Norm of x: %2.4f" % norm_x_2
print "Normalized x MLlib:\n%s" % normalized_x_mllib
print "2-Norm of normalized_x_mllib: %2.4f" % np.linalg.norm(normalized_x_mllib)

File: src/main/scala/org/hdwyl/MovieStats.scala
package org.hdwyl

import java.util.Calendar

import breeze.linalg._
import org.apache.log4j.Logger
import org.apache.spark.{SparkConf, SparkContext}

import scala.util.matching.Regex

/**
  * Created by wangyanl on 2019/6/14.
  */
object MovieStats {

  @transient lazy val logger = Logger.getLogger(this.getClass)

  def main(args: Array[String]) {
    val conf = new SparkConf()
    val sc = new SparkContext(conf)
    // 读取数据集
    val movieData = sc.textFile("hdfs://PATH/ml-100k/u.item")
    // 输出第1行数据
    println(movieData.first())
    // 输出前k行数据
    println(movieData.take(10))

    // 统计电影数量
    val numMovies = movieData.count()
    // Movies: 1682
    println("Movies: %d".format(numMovies))

    val movieFields = movieData.map(line => line.split("|"))
    // 提取出电影的年份信息
    val years = movieFields.map(e => e(2)).map(x => convertYear(x))
    // 过滤掉1900年(即未记录年份)的电影数据
    val filteredYears = years.filter { year => year != 1900 }
    // 计算得到电影的年龄
    val thisYear = Calendar.getInstance().get(Calendar.YEAR)
    val movieAges = filteredYears.map(year => thisYear - year).countByValue()
    movieAges.foreach(println)

    val meanYear = filteredYears.sum() / filteredYears.count()
    // 非1900年的全部年份的中位数值
    val medianYear = Utils.getMedian(filteredYears)
    println(years.filter(year => year == 1900))
    val yearsPreProcessed = years.map(year => if (year == 1900) medianYear else year)
    println(yearsPreProcessed.filter(year => year == 1900))
    // Mean year of release: 1989
    println("Mean year of release: %d".format(meanYear))
    // Median year of release: 1995
    println("Median year of release: %d".format(medianYear))
    // Index of '1900' before assigning median: List(266)
    println("Index of '1900' before assigning median: %s".format(years.zipWithIndex.collect().find(e => e._1 == 1900).map(e => e._2).toList))
    // Index of '1900' after assigning median: List()
    println("Index of '1900' after assigning median: %s".format(yearsPreProcessed.zipWithIndex.collect().find(e => e._1 == 1900).map(e => e._2).toList))

    val rawTitles = movieFields.map(e => e(1))
    val movieTitles = rawTitles.map(title => extractTitle(title))
    // 下面用简单空白分词法将标题分词为词
    val titleTerms = movieTitles.map(title => title.split(" "))
    // 下面取回所有可能的词，以便构建一个词到序号的映射字典
    val pattern = new Regex("[,\\(\\):]")
    val allTermsWithIndex = titleTerms.flatMap(e => e).map(e => pattern replaceAllIn (e, "")).distinct.zipWithIndex.collectAsMap()
    // Total number of terms: 2457
    println("Total number of terms: %d".format(allTermsWithIndex.size))
    // Index of term 'Dead': 19
    println("Index of term 'Dead': %d".format(allTermsWithIndex.get("Dead").get))
    // Index of term 'Rooms': 4
    println("Index of term 'Rooms': %d".format(allTermsWithIndex.get("Rooms").get))

    val allTermsWithIndexBcast = sc.broadcast(allTermsWithIndex)
    val termVectors = titleTerms.map(terms => createVector(terms.toList, allTermsWithIndexBcast.value.toMap))
    titleTerms.take(10).map(term => term.mkString(" ")).foreach(println)
    termVectors.take(10).foreach(println)

    sc.stop()
  }

  def convertYear(x: String): Int = {
    try {
      return Integer.parseInt(x.substring(x.length() - 4))
    } catch {
      case ex: NumberFormatException => {
        return 1900 // 若数据缺失年份则将其年份设为1900
      }
      case ex: StringIndexOutOfBoundsException => {
        println(x)
        return 1900 // 若数据缺失年份则将其年份设为1900
      }
    }
  }

  def extractTitle(raw: String): String = {
    // 该表达式找寻括号之间的非单词（数字）
    val pattern = new Regex("\\(\\d+\\)")
    return (pattern replaceFirstIn(raw, "")).trim
  }

  // 该函数输入一个词列表，并用k之1编码类似的方式将其编码为一个稀疏向量
  def createVector(terms: List[String], termDict: Map[String, Long]): DenseMatrix[Int] = {
    import breeze.linalg.DenseMatrix._
    val numTerms = termDict.size
    val x = DenseMatrix.zeros[Int](1, numTerms)
    for (term <- terms) {
      if (termDict.contains(term)) {
        val idx = termDict.get(term).get.toInt
        x(::, idx) := 1
      }
    }
    return x
  }
}
